### ISSUE TITLE
Any Ideas on How to Write or Adapt Keras Modules for Recursive Neural Networks.

### ISSUE THREAD

**Issue by simonhughes22** [fchollet/keras#352]
Note recursive not recurrent. The sort developed by the Stanford group, in particular Richard Socher (see richardsocher.org) of MetaMind, such as the recursive auto-encoder, the tree-based networks and the Tree-LSTM?

**Comment by fchollet** [comment]
That sounds like it would definitely be possible. What have you tried so far? What task are you working on?

**Comment by simonhughes22** [comment]
I haven't :). I need to learn more theano first. I presume we'd need a way to represent a tree structure somehow, or a way to reconstruct the tree from a flattened list. I am working on text classification tasks, specifically the rather challenging task of determining causality in a sentence. Which can essentially be framed as either a word tagging model or a sentence classification model (the sentence labels are what we care about, but we have word level tags). I am not sure a recursive NNet will be better given some recent strong results using convolutional nets, but I am interested in trying it out.

**Comment by wuxianliang** [comment]
https://github.com/stanfordnlp/treelstm
Socher's Tree-LSTM is realized by Torch. So I think it is a good starting point to translate the project into Keras.

**Comment by wuxianliang** [comment]
an example of treelstm in theano
https://github.com/Azrael1/Seq-Gen/blob/master/models/prelim_models/model2.0/treelstm.py

**Comment by ankitp94** [comment]
@wuxianliang  Thank you for the tree-lstm link.
Is there any example/readme/tutorial for tree-lstm theano code (from pre-processing the dataset to prediction) ?

**Comment by pasky** [comment]
Maybe porting the NNBlock's Theano implementation of ReNNs might be a good starting point. Feels to me (as a Keras newbie) that the trickiest part will be just figuring out how to represent the input tree in a Keras-friendly format.

**Comment by Imorton-zd** [comment]
Hi, @simonhughes22 , have you  implemented the auto-encoder by keras? I want to use auto-encoder for texts classification, would you give me some suggestions?

**Comment by mardan** [comment]
@ankitp94 Hi! Did you make tree-lstm work for you? I am looking for theano tree-lstm example dealing with dependency structure of sentence.

**Comment by tsterbak** [comment]
Has one of you guys made any progress in this direction?

**Comment by arunpatala** [comment]
Can anybody give me some pointers to implement this. Should we use the graph module to implement this (any examples)? As the tree structure is variable among different sentences, how to provide input and do batch processing ? 

I am going through the torch-lstm example. They seem to process each tree separately (no batches). And they seem to create a new graph to process each tree, while sharing the parameters between the different graphs. So do the following steps seems reasonable?

1. Pass dictionary based input for the graph to pass the tree (will this work?)
2. For each tree, create the graph reflecting the tree. 
3. Use shared parameters (or module) every time for the recursive operation.
4. Call fit on that model for that graph on that single tree.

**Comment by shaktisd** [comment]
Any progress on developing Tree LSTM in keras ?

**Comment by pengpaiSH** [comment]
+1 Any progress on developing Tree LSTM in keras ?

**Comment by deepanwayx** [comment]
Any update on this topic?

**Comment by baregawi** [comment]
One possibility that works decently for some language problems I've tried is to serialize the input tree in the order of the recursion you are aiming for and then using a normal LSTM along that sequence. In terms of the final computational graph, I'm not sure there is any difference between that and what a tree LSTM does other than that you have to pad the sequence for variable size inputs which is not an issue due to the masking layer.
